### PR TITLE
chore(core): formalize vocabulary registries for entity_type, source_type, relation_type

### LINKED ISSUE #199: chore(core): formalize vocabulary registries (entity_type, source_type, relation_type)
## Goal

Create centralized TypeScript registries for `entity_type`, `source_type`, and
`relation_type` values. All ingesters and adapters import from the registry rather than
inlining string literals.

## Problem

String values are scattered across the codebase with no single source of truth:

- `entity_type` values used today: `person`, `service`, `module`, `issue`, `commit`,
  `pull_request` — hardcoded per ingester.
- `source_type` values: `git`, `git_commit`, `github_pr`, `github_issue`, `manual`,
  `document`, `source`, `github`. **Note the asymmetry** — `ingestion_runs.source_type`
  uses `"github"` while episodes use `"github_pr"`/`"github_issue"`. Invisible until
  the cross-source ref resolver (#197) needed to match against these values.
- `relation_type` values: `authored_by`, `reviewed_by`, `co_changes_with`,
  `likely_owner_of`, `references`, `depends_on`, `calls`, `uses`, `maintains`, etc.

With Gerrit, Buganizer, Google Docs, GitLab, Jira, Linear adapters landing, the problem
compounds. A typo (`gerrit_change` vs `gerrit-change`) silently breaks cross-ref
resolution and nothing catches it.

**Scale of refactor:** 24 string-literal sites across 5 files in `src/ingest/`
(`git.ts`, `adapters/github.ts`, `source/index.ts`, `markdown.ts`, `text.ts`) —
bounded.

### Type surface

Each registry follows a uniform pattern — `const` object + derived union type:

```ts
// packages/engram-core/src/vocab/entity-types.ts
export const ENTITY_TYPES = {
  PERSON: 'person',
  MODULE: 'module',
  SERVICE: 'service',
  ISSUE: 'issue',
  COMMIT: 'commit',
  PULL_REQUEST: 'pull_request',
  // ...
} as const;

export type EntityType = typeof ENTITY_TYPES[keyof typeof ENTITY_TYPES];
```

Gives both a runtime constant (for writes and iteration) and a compile-time union (for
function signatures). Matches the `AuthCredential` union style introduced in #200.

### Enforcement layer — runtime only, no SQL CHECK

The SQLite schema stays permissive (`TEXT NOT NULL` as today). Enforcement happens in
two places:

1. **Compile time:** ingester and adapter code imports from the registry; TS rejects
   unknown literals at the type boundary.
2. **Runtime:** `verifyGraph` gains a strict mode that flags any row whose type value
   isn't in the applicable registry as a `warning`.

No format-version bump, no migration for existing `.engram` files. Consistent with the
existing pattern: schema is permissive, `verifyGraph` is authoritative.

### Two source_type registries

The `ingestion_runs.source_type` vs `episodes.source_type` asymmetry is a real semantic
distinction — one ingestion run can emit multiple episode kinds. Formalize with two
registries and an explicit mapping:

```ts
// packages/engram-core/src/vocab/source-types.ts

/** Values used in ingestion_runs.source_type — identifies the ingestion pass. */
export const INGESTION_SOURCE_TYPES = {
  GIT: 'git',
  GITHUB: 'github',
  SOURCE: 'source',
  // ...
} as const;

/** Values used in episodes.source_type — identifies the episode kind. */
export const EPISODE_SOURCE_TYPES = {
  GIT_COMMIT: 'git_commit',
  GITHUB_PR: 'github_pr',
  GITHUB_ISSUE: 'github_issue',
  MANUAL: 'manual',
  DOCUMENT: 'document',
  SOURCE_FILE: 'source_file',
  // ...
} as const;

/** Which episode kinds an ingestion source emits. */
export const INGESTION_TO_EPISODE_SOURCES = {
  [INGESTION_SOURCE_TYPES.GIT]: [EPISODE_SOURCE_TYPES.GIT_COMMIT],
  [INGESTION_SOURCE_TYPES.GITHUB]: [
    EPISODE_SOURCE_TYPES.GITHUB_PR,
    EPISODE_SOURCE_TYPES.GITHUB_ISSUE,
  ],
  // ...
} as const;
```

No data migration — both columns keep their current semantics.

### Closed registry for v0.1

All registry values ship in `packages/engram-core/src/vocab/`. Third-party adapters
cannot add values at load time in v0.1 — revisit when the plugin-loading architecture
(#204) lands. If #204 concludes plugins need dynamic type registration, add an escape
hatch then.

## Scope

1. Create `packages/engram-core/src/vocab/` with:
   - `entity-types.ts` — `ENTITY_TYPES` + `EntityType`
   - `source-types.ts` — `INGESTION_SOURCE_TYPES`, `EPISODE_SOURCE_TYPES`,
     `INGESTION_TO_EPISODE_SOURCES` + derived types
   - `relation-types.ts` — `RELATION_TYPES` + `RelationType`
   - `index.ts` — barrel exports
2. Wire into `packages/engram-core/src/index.ts` public API surface.
3. Refactor `src/ingest/` — replace the 24 string-literal sites (git.ts,
   adapters/github.ts, source/index.ts, markdown.ts, text.ts) with registry imports.
4. Extend `verifyGraph` (`format/verify.ts`) with an optional strict mode
   (`verifyGraph(graph, { strict: true })`) that flags unknown-vocab rows as
   `warning`-severity violations.

## Out of scope

- **SQL CHECK constraints** — deferred. Would require a format version bump and
  migration for existing `.engram` files. Runtime enforcement via `verifyGraph` is
  sufficient for v0.1.
- **Adapter-defined / plugin-contributed values** — closed registry only. Revisit
  when #204 lands.
- **Rewriting test fixtures** — tests may continue using string literals where doing
  so is clearer than importing registry constants.

## Edge cases

- **Legacy `.engram` files.** Files written before the registry landed may contain
  values absent from the registry (e.g. a retired `entity_type`). `verifyGraph` strict
  mode reports these as warnings, not errors. Normal (non-strict) `verifyGraph` ignores
  them entirely.
- **In-flight adapters.** Adapters written against older versions of the package still
  work — the registry change is type-level and runtime-warning-level only; no
  behavior change for non-strict callers.
- **Registry drift over time.** Retired values (e.g. if `service` stops being used)
  stay in the registry as `@deprecated` until a major version — removing a value is
  breaking.
- **Cross-registry typos.** `relation_type: 'authored_by'` vs `entity_type: 'authored_by'`
  — the three registries are distinct types; mixing is a TS compile error.
- **Unknown values in strict-mode verification.** Warning message includes the offending
  value, the row id, and the applicable registry name so users can see whether to add
  the value or fix the data.

## Acceptance criteria

- [ ] `packages/engram-core/src/vocab/{entity-types,source-types,relation-types,index}.ts`
      exist with the `as const` + derived-union pattern
- [ ] Source-type split: separate `INGESTION_SOURCE_TYPES` and `EPISODE_SOURCE_TYPES`
      registries + `INGESTION_TO_EPISODE_SOURCES` mapping — asymmetry is documented
      and machine-readable
- [ ] `packages/engram-core/src/index.ts` re-exports the vocab module as part of the
      public API
- [ ] Zero string literals for these three columns remain in `src/ingest/`
      (`git.ts`, `adapters/github.ts`, `source/index.ts`, `markdown.ts`, `text.ts`);
      tests may still use literals where clearer
- [ ] `verifyGraph(graph, { strict: true })` added; flags unknown-vocab rows as
      `warning`-severity violations with a message naming the registry and offending
      value
- [ ] Normal `verifyGraph(graph)` behavior unchanged (no new warnings for existing
      callers)
- [ ] New adapters land after this issue (#191, #196, etc.) consume the registry from
      day one
- [ ] Plugin-declared vocab extensions (per #204) are additive via manifest only and
      merge into the same in-memory registry; no runtime stringing. The built-in
      registry shipped in this issue remains closed — `vocab/` is the single source of
      truth for built-in values, and plugins extend it declaratively at load time.

## Tests

- [ ] Type-level test (compile-only, e.g. `expectType` or `// @ts-expect-error`): bad
      literal assignment to `entity_type` fails to compile
- [ ] Unit test for `verifyGraph({ strict: true })`: seed a graph with a known-bad
      `entity_type` value, assert a `warning` violation is surfaced with the correct
      message
- [ ] Unit test: strict mode produces no violations on a graph where all rows use
      registry values
- [ ] Unit test: non-strict `verifyGraph` produces no new warnings on a legacy graph
      containing off-registry values (backward-compat assertion)
- [ ] Integration test: `INGESTION_TO_EPISODE_SOURCES` mapping is consistent with what
      each existing ingester actually emits (iterate + spot-check against fresh
      ingestion output)

## Documentation

- [ ] `docs/internal/specs/vocabulary.md` (or a section in an existing spec)
      enumerates all three registries, the ingestion-vs-episode source_type
      convention, and the "closed for v0.1" policy
- [ ] CLAUDE.md "Data Model" section updated: replace inline type lists with a pointer
      to the vocabulary spec and the registry module
- [ ] CLAUDE.md "Ingestion Architecture" section: note that new adapters must import
      from `packages/engram-core/src/vocab/` for the three type columns
- [ ] CLAUDE.md "Key Files" row for `packages/engram-core/src/vocab/` and
      `packages/engram-core/src/format/verify.ts` (strict mode)
- [ ] JSDoc on each registry constant explaining the naming convention (snake_case
      values) and when to add a new value (with a reference to this issue's design
      decisions)

## Blocks

- **#200** — EnrichmentAdapter contract v2 declares #199 as its blocker; v2 adapters
  must consume the registry from day one
- **#197** — cross-source reference edge resolver needs stable values to match against
- **#191** (Gerrit), **#193** (GitLab), **#194** (Jira), **#195** (Linear),
  **#196** (Google Docs), **#198** (Google Drive), **#192** (Buganizer spike)

## Related

- **#204** — plugin-loading architecture may reopen the "closed registry" decision
- **#205** — CLI v2 consumer; no direct dependency but will surface registry-driven
  errors to users

### PR DESCRIPTION
**Pull request by rnwolfe**
## Summary

Closes #199

- Introduces centralized vocabulary registries (`ENTITY_TYPES`, `EPISODE_SOURCE_TYPES`, `INGESTION_SOURCE_TYPES`, `RELATION_TYPES`) in `packages/engram-core/src/vocab/`
- Replaces 24+ string literals across `git.ts`, `github.ts`, `source/index.ts`, `markdown.ts`, and `text.ts` with registry constants
- Formalizes the two-registry pattern for source types: `INGESTION_SOURCE_TYPES` (for ingestion runs) vs `EPISODE_SOURCE_TYPES` (for episodes), with an `INGESTION_TO_EPISODE_SOURCES` mapping
- Adds `verifyGraph({ strict: true })` support — flags unknown-vocab rows as warnings in tests
- Adds `docs/internal/specs/vocabulary.md` spec documenting all registered values, the closed-registry policy, and the procedure for adding new values

## Acceptance Criteria

- [x] `ENTITY_TYPES`, `EPISODE_SOURCE_TYPES`, `INGESTION_SOURCE_TYPES`, `RELATION_TYPES` constants defined in `src/vocab/`
- [x] All ingest adapters import from `src/vocab/index.js` — no inline string literals for entity_type, source_type, or relation_type
- [x] `verifyGraph({ strict: true })` validates that stored values are in the registered set
- [x] Tests added for strict-mode vocab checks (5 new tests in `verify.test.ts`)
- [x] Vocabulary spec document added at `docs/internal/specs/vocabulary.md`
- [x] Build passes, 870/873 tests pass (3 pre-existing failures unrelated to this PR)

## Test plan

- [ ] `bun test` passes
- [ ] `bun run build` passes
- [ ] `bun run lint` passes

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: packages/engram-core/src/ingest/git.ts [hunk 1]
```diff
@@ -168,7 +174,7 @@ function getLastCursor(graph: EngramGraph, sourceScope: string): string | null {
   const row = graph.db
     .query<{ cursor: string | null }, [string]>(
       `SELECT cursor FROM ingestion_runs
-       WHERE source_type = 'git_commit' AND source_scope = ? AND status = 'completed'
+       WHERE source_type = '${INGESTION_SOURCE_TYPES.GIT}' AND source_scope = ? AND status = 'completed'
        ORDER BY completed_at DESC LIMIT 1`,
```
Comment (rnwolfe) [inline packages/engram-core/src/ingest/git.ts:178]: **[medium] Template literal embeds constant into SQL — prefer parameterized query for consistency**

```ts
`SELECT cursor FROM ingestion_runs
 WHERE source_type = '${INGESTION_SOURCE_TYPES.GIT}' AND source_scope = ? AND status = 'completed'`
```

`INGESTION_SOURCE_TYPES.GIT` is a compile-time constant (`"git"`) so there is no injection risk here. However, using a template literal to splice a value into SQL is stylistically inconsistent with every other query in this codebase, which uses `?` placeholders. If the constant ever becomes user-supplied or is renamed to a value containing a single quote, this will silently break. Prefer:

```ts
`SELECT cursor FROM ingestion_runs
 WHERE source_type = ? AND source_scope = ? AND status = 'completed'`
```
and pass `INGESTION_SOURCE_TYPES.GIT` as a positional parameter.

File: packages/engram-core/src/ingest/source/index.ts [hunk 2]
```diff
@@ -153,7 +159,7 @@ function findActiveEpisodeForPath(
     graph.db
       .query<{ id: string; source_ref: string }, [number, string]>(
         `SELECT id, source_ref FROM episodes
-         WHERE source_type = 'source'
+         WHERE source_type = '${EPISODE_SOURCE_TYPES.SOURCE_FILE}'
            AND status = 'active'
```
Comment (rnwolfe) [inline packages/engram-core/src/ingest/source/index.ts:163]: **[medium] Template literal SQL uses `EPISODE_SOURCE_TYPES.SOURCE_FILE` while episodes are written with `SOURCE_TYPE` (= `INGESTION_SOURCE_TYPES.SOURCE`)**

Both constants evaluate to the string `"source"`, so the queries are functionally correct today. But using two different registry keys to refer to the same column value in the same file is semantically misleading: `SOURCE_TYPE` is `INGESTION_SOURCE_TYPES.SOURCE` (an ingestion-run key), while the SQL queries here read from `episodes.source_type` and correctly use `EPISODE_SOURCE_TYPES.SOURCE_FILE`.

The fix is to replace `SOURCE_TYPE` in the episode-write call (line ~303) with `EPISODE_SOURCE_TYPES.SOURCE_FILE`, and similarly update the idempotency fast-path query (line ~237) so every reference to `episodes.source_type` uses the episode registry:

```ts
// episode write
source_type: EPISODE_SOURCE_TYPES.SOURCE_FILE,

// idempotency query
.get(EPISODE_SOURCE_TYPES.SOURCE_FILE, sourceRef);
```

`SOURCE_TYPE` / `INGESTION_SOURCE_TYPES.SOURCE` should then only appear in the `ingestion_runs` insert.

File: packages/engram-core/src/ingest/source/index.ts [hunk 3]
```diff
@@ -535,7 +541,7 @@ export async function ingestSource(
     const sweepCandidates = graph.db
       .query<{ id: string; source_ref: string }, [string]>(
         `SELECT id, source_ref FROM episodes
-         WHERE source_type = 'source'
+         WHERE source_type = '${EPISODE_SOURCE_TYPES.SOURCE_FILE}'
            AND status = 'active'
```
Comment (rnwolfe) [inline packages/engram-core/src/ingest/source/index.ts:545]: **[minor] Same template-literal-in-SQL style issue as `git.ts` line 177**

These three SQL fragments (`findActiveEpisodeForPath`, and the two cleanup queries at the bottom of `ingestSource`) all splice `EPISODE_SOURCE_TYPES.SOURCE_FILE` via a template literal instead of using a `?` placeholder. Prefer parameterized queries for the same consistency reasons noted in `git.ts`.

**Review by rnwolfe — COMMENTED** [review]
Review by forge-loop sub-agent

File: packages/engram-core/test/format/verify.test.ts [hunk 4]
```diff
@@ -15,6 +15,7 @@ import {
   createGraph,
   verifyGraph,
 } from "../../src/index.js";
+import { EPISODE_SOURCE_TYPES } from "../../src/vocab/index.js";
```
Comment (rnwolfe) [inline packages/engram-core/test/format/verify.test.ts:19]: **[minor] `makeEpisode` and `makeEdge` test helpers still inline string literals — sets a bad example in the test added by this PR**

The existing `makeEpisode`, `makeEntity`, and `makeEdge` helpers (unchanged lines above) use hardcoded literals `"manual"`, `"person"`, and `"knows"`. This PR introduces a new test suite explicitly about the vocab registry, and the very test that verifies "existing registry values produce no warnings" (the last test) also inlines `entity_type: "person"` (line 102 in the diff) instead of `ENTITY_TYPES.PERSON`.

For the new tests, at minimum the backward-compat test should be fully consistent:

```ts
addEntity(
  graph,
  { canonical_name: "alice@example.com", entity_type: ENTITY_TYPES.PERSON },
  [{ episode_id: ep.id, extractor: "test", confidence: 1.0 }],
);
```

Updating the helper functions is a pre-existing issue, but the new test adding `entity_type: "person"` inline is a new regression against the policy this very PR is establishing.

File: packages/engram-core/test/format/verify.test.ts [hunk 5]
```diff
@@ -484,3 +485,103 @@ describe("VerifyResult.valid", () => {
...
+      (v) => v.check === "checkVocab" && v.entity_or_edge_id === entity.id,
+    );
+    expect(v).toBeDefined();
+    expect(v?.severity).toBe("warning");
+    expect(v?.message).toContain("legacy_thing");
+    expect(v?.message).toContain("ENTITY_TYPES");
+  });
+
+  test("flags unknown episode source_type as warning in strict mode", () => {
+    const ep = makeEpisode(graph);
```
Comment (rnwolfe) [inline packages/engram-core/test/format/verify.test.ts:523]: **[medium] Missing test: no coverage for unknown `ingestion_runs.source_type` in strict mode**

`checkVocab` validates four columns:
1. `entities.entity_type` ✅ tested (line 29)
2. `episodes.source_type` ✅ tested (this test, line 46)
3. `ingestion_runs.source_type` ❌ not tested
4. `edges.relation_type` ✅ tested (line 61)

There is no test that inserts an ingestion run with an unknown `source_type` and asserts that `checkVocab` flags it. Since `ingestion_runs` is only populated by ingest functions (not by `addEpisode`/`addEntity`), the missing case may be harder to trigger via the public API, but a direct `graph.db.run()` insert — the same pattern used by the other tests — would cover it. Example:

```ts
test("flags unknown ingestion_run source_type as warning in strict mode", () => {
  const runId = ulid();
  graph.db.run(
    "INSERT INTO ingestion_runs (id, source_type, source_scope, started_at, extractor_version, episodes_created, entities_created, edges_created, status) VALUES (?, ?, ?, ?, ?, 0, 0, 0, ?)",
    [runId, "unknown_ingestor", "/some/repo", new Date().toISOString(), "0.0.0", "completed"],
  );
  const result = verifyGraph(graph, { strict: true });
  const v = result.violations.find(
    (v) => v.check === "checkVocab" && v.entity_or_edge_id === runId,
  );
  expect(v).toBeDefined();
  expect(v?.severity).toBe("warning");
});
```

**Review by rnwolfe — COMMENTED** [review]
Review by forge-loop sub-agent (part 2)

**Comment by rnwolfe** [comment]
All review findings addressed in commit e2c3b05:

- **Medium #1** (git.ts template literal SQL): Converted `getLastCursor` query to parameterized `?` placeholder
- **Medium #2** (source/index.ts semantic inconsistency): All episode queries now use `EPISODE_SOURCE_TYPES.SOURCE_FILE` consistently; removed mixed use of `SOURCE_TYPE` (which was the ingestion-run constant)
- **Minor #3 + #4** (template literal SQL in source/index.ts): All 3 sweep queries converted to parameterized queries
- **Minor #5** (inline 'person' string): Changed to `ENTITY_TYPES.PERSON` in backward-compat test
- **Missing coverage**: Added test for unknown `ingestion_runs.source_type` in strict mode

Build passes, 872/874 tests pass (2 pre-existing unrelated failures).
